Create a CLI

rgreinho opened the pull request:
Creates a CLI for OpenAlchemy. The first command available allows the
users to build a package for the models and a distributable archive.

rgreinho commented:
I usually do not test my CLIs, because they are simply wrappers around backend functions which as well tested. I was not sure what your take on this was so I did not remove it from the coverage. 

If you would like me to test it, would you happen to have some ideas/feedback to do so? I know `click` has some mechanism for [testing](https://click.palletsprojects.com/en/7.x/testing/), but at the end of he day it is only to validate the command arguments 🤷

rgreinho commented:
I created another draft PR (https://github.com/jdkandersson/OpenAlchemy/pull/202), which is add a new subcommand, but it is more for discussion for now. 

Adding the other command was useful to see how to refactor and test the code. I would still like to submit 2 different PRs, one for each command, but I thought it would be a good idea to discuss the implementation beforehand.

rgreinho commented:
In this PR, I removed the `.python-versions` file, because it was preventing me from starting vscode, and even creating my venv.

I figured out that since there are 3 python versions specified in the file, **ALL the versions MUST** be installed with `pyenv`. Maybe only putting the latest version in this file should be enough then?

jdkandersson commented:
> In this PR, I removed the `.python-versions` file, because it was preventing me from starting vscode, and even creating my venv.
> 
> I figured out that since there are 3 python versions specified in the file, **ALL the versions MUST** be installed with `pyenv`. Maybe only putting the latest version in this file should be enough then?

I don't mind the removal, I believe it was required for tox. I should remove the tax tests as I haven't been using them and instead only testing the remainder on the CI.

jdkandersson commented:
> I usually do not test my CLIs, because they are simply wrappers around backend functions which as well tested. I was not sure what your take on this was so I did not remove it from the coverage.
> 
> If you would like me to test it, would you happen to have some ideas/feedback to do so? I know `click` has some mechanism for [testing](https://click.palletsprojects.com/en/7.x/testing/), but at the end of the day it is only to validate the command arguments which is done by the CLI parser anyway. 🤷

Is it possible to test it in a similar way as was done with the build command? I.e. you just create a subprocess and try to run the command?

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+    build_parser.add_argument(
+        "specfile", type=str, help="specify the specification file"
+    )
+    build_parser.add_argument("name", type=str, help="specify the package name")
```
Should we add an example? e.g. `"specify the package name, e.g. 'my-package'"`

rgreinho replied:
I don't think we should add examples in the CLI. It would get cluttered very quick. Besides it wraps at 80 characters. 

Also, looking at other CLIs (e.g. `docker`, `git`), they don't provide examples on the help screen (but they do in the documentation).

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+        "specfile", type=str, help="specify the specification file"
+    )
+    build_parser.add_argument("name", type=str, help="specify the package name")
+    build_parser.add_argument("output", type=str, help="specify the output directory")
```
Like above, an example would be great here

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+    )
+    build_parser.add_argument("name", type=str, help="specify the package name")
+    build_parser.add_argument("output", type=str, help="specify the output directory")
+    build_parser.add_argument(
```
Is it possible to do both?

rgreinho replied:
Addressed ✅

jdkandersson replied:
Do you think there is any value in calling out that it is optional? I will take your lead on that

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+        logging.error("Cannot perform the operation: %s.", exc)
+
+
+def build_application_parser():
```
Please add type hints (e.g. `def build_application_parser() -> None:` if the function returns nothing)

rgreinho replied:
Addressed ✅

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+    return parser.parse_args()
+
+
+def build(args):
```
Please also add type hints

rgreinho replied:
Addressed ✅

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+
+
+def build(args):
+    """Define the build subcommand."""
```
Please describe the argument and return value in the docstring

rgreinho replied:
Addressed ✅

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+    # Check the specfile.
+    specfile = pathlib.Path(args.specfile)
+    if not specfile.exists():
+        raise ValueError(f"the specfile '{specfile}' cannot be found")
```
Please use exceptions inheriting from the `open_alchemy.exceptions.BaseException`.

rgreinho replied:
Addressed ✅

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+    }
+    ext = specfile.suffix.lower()
+    if ext not in builders.keys():
+        raise ValueError(
```
The same feedback here, please use exceptions inheriting from the `open_alchemy.exceptions.BaseError`.

rgreinho replied:
Addressed ✅

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+    args = build_application_parser()
+    try:
+        args.func(args)
+    except Exception as exc:  # pylint: disable=broad-except
```
Is it reasonable to articulate the possible exceptions here? OpenAlchemy can only raise exceptions inheriting from `open_alchemy.exceptions.BaseError` and has been designed to handle any possible errors that could occur during execution gracefully. I might have missed some, they would be good to add and tell users what to change to avoid the error.

rgreinho replied:
Addressed ✅

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,82 @@
+    if ext not in builders.keys():
+        raise ValueError(
+            f"specification format not supported: '{args.specfile}' "
+            "(should be json or yaml)"
```
It would be excellent to describe what was received versus what was expected. I.e. `"received extension <extension>, supported extensions: .yaml, .yml, .json (case insensitive)"`

rgreinho replied:
Addressed ✅

jdkandersson commented:
Please add documentation :smile:

Please also update the readme with the additional feature.

rgreinho commented:
> > In this PR, I removed the `.python-versions` file, because it was preventing me from starting vscode, and even creating my venv.
> > I figured out that since there are 3 python versions specified in the file, **ALL the versions MUST** be installed with `pyenv`. Maybe only putting the latest version in this file should be enough then?
> 
> I don't mind the removal, I believe it was required for tox. I should remove the tox tests as I haven't been using them and instead only testing the earlier versions and other platforms on CI.

That works for me. I'll remove the file. 👍

rgreinho commented:
> > I usually do not test my CLIs, because they are simply wrappers around backend functions which as well tested. I was not sure what your take on this was so I did not remove it from the coverage.
> > If you would like me to test it, would you happen to have some ideas/feedback to do so? I know `click` has some mechanism for [testing](https://click.palletsprojects.com/en/7.x/testing/), but at the end of the day it is only to validate the command arguments which is done by the CLI parser anyway. 🤷
> 
> Is it possible to test it in a similar way as was done with the build command? I.e. you just create a subprocess and try to run the command?
> 
> The argument for testing the CLI is to ensure that, if someone changes the interface of the functions exposed by the CLI, that those tests fail and let the person know to also update the CLI.
> 
> Understanding this better now, It would be really good to exercise the CLI Python functions. A stretch would be to also execute example CLI commands as described above, they can help serve as examples. I have never worked with CLIs before but want to get better. If you would like, I'm happy to merge this into a feature branch and write the tests on your behalf. I'm equally happy for you to write those tests.

Since argparse uses `sys.argv` as input, I just set it to the value I want in the test. We could also achieve the same thing using the `run()` function, but I think it is simpler this way.

The `test_main()` test is parametrized and allows to call any command we want with any combination of arguments. If you want to add more test cases, that would be there.

rgreinho commented:
I believe I addressed all your comments. Except for the documentation.

The README starts to get really long. What would you think of created a dedicated page for the CLI? Or maybe giving a small example in the README, but the having an "Advance usage page" (something more or less like https://www.scrapd.org/usage.html)?

We could also use an extension which generates parts of the CLI documentation based on the argparse: https://sphinx-argparse.readthedocs.io/en/latest/.

Let me know what your thoughts are regarding this part and I'll start working on the documentation.

rgreinho commented:
Oh you switched the CI to GitHub Actions! Very cool! However for some reasons which seem unrelated to my PR, the CI is very unhappy with this patch 🤔

jdkandersson commented:
> Oh you switched the CI to GitHub Actions! Very cool! However for some reasons which seem unrelated to my PR, the CI is very unhappy with this patch 🤔

I think it is because some of them are failing so I assume it cancels the rest once it detects a failure.

jdkandersson commented:
> > > I usually do not test my CLIs, because they are simply wrappers around backend functions which as well tested. I was not sure what your take on this was so I did not remove it from the coverage.
> > > If you would like me to test it, would you happen to have some ideas/feedback to do so? I know `click` has some mechanism for [testing](https://click.palletsprojects.com/en/7.x/testing/), but at the end of the day it is only to validate the command arguments which is done by the CLI parser anyway. 🤷
> > 
> > 
> > Is it possible to test it in a similar way as was done with the build command? I.e. you just create a subprocess and try to run the command?
> > The argument for testing the CLI is to ensure that, if someone changes the interface of the functions exposed by the CLI, that those tests fail and let the person know to also update the CLI.
> > Understanding this better now, It would be really good to exercise the CLI Python functions. A stretch would be to also execute example CLI commands as described above, they can help serve as examples. I have never worked with CLIs before but want to get better. If you would like, I'm happy to merge this into a feature branch and write the tests on your behalf. I'm equally happy for you to write those tests.
> 
> Since argparse uses `sys.argv` as input, I just set it to the value I want in the test. We could also achieve the same thing using the `run()` function, but I think it is simpler this way.
> 
> The `test_main()` test is parametrized and allows to call any command we want with any combination of arguments. If you want to add more test cases, that would be there.

Yeh I think that is ok. It would still be good to have 1 test using `run` just to confirm any assumptions we're making about `sys.argv`.

jdkandersson commented:
> I believe I addressed all your comments. Except for the documentation.
> 
> The README starts to get really long. What would you think of created a dedicated page for the CLI? Or maybe giving a small example in the README, but the having an "Advance usage page" (something more or less like https://www.scrapd.org/usage.html)?
> 
> We could also use an extension which generates parts of the CLI documentation based on the argparse: https://sphinx-argparse.readthedocs.io/en/latest/.
> 
> Let me know what your thoughts are regarding this part and I'll start working on the documentation.

I think that should be ok, I'm curios to see what they look like. It is important that the docs can still be generated locally (using `make html`) and it would be good to know if there are any impacts to the current read the docs integration.

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,105 @@
+    """
+    # Ensure the file exists.
+    if not specfile.exists():
+        raise exceptions.BaseError(f"the specfile '{specfile}' cannot be found")
```
My apologies, I wasn't specific enough. I wouldn't expect `BaseError` to ever get raised, I would expect a more specific error to be raised that inherits from `BaseError`. `BaseError` is designed to go into a `try ... except ...` block to catch any OpenAlchemy error. I don't think there is a good error at the moment, perhaps a `CliError` can be defined here?

jdkandersson commented on open_alchemy/cli.py:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    # Build the package.
+    builder = builders.get(specfile.suffix.lower())
```
Would this now raise a `KeyError` if the format is not supported? I liked the previous explicit check and exception which explains what is wrong, a generic `KeyError` might be hard for people to understand.

jdkandersson replied:
Oops, did not spot the validate spec file.

jdkandersson commented on setup.cfg:
```diff
@@ -4,26 +4,27 @@ universal = 1
-    artifacts
-    validation
-    build
+  app
```
I'm interested in this, did you manually re-order it?

rgreinho replied:
Nope, I used https://www.online-utility.org/text/sort.jsp.

jdkandersson replied:
Ah nice, might be splitting hairs, another part of this file uses a tab width of 4 😄

jdkandersson commented on tests/test_cli.py:
```diff
@@ -0,0 +1,102 @@
+    WHEN validate_specfile is called
+    THEN nothing happens
+    """
+    spec = tmp_path / specfile
```
I noted this in the `CONTRIBUTING.md` file, please keep a blank line between the code that implements GIVEN, WHEN and THEN. In this case line 28 and 29 to me look like GIVEN, and line 30 is a WHEN and there is no THEN since the expectation is that no error is raised

rgreinho replied:
Copy that!

jdkandersson commented on tests/test_cli.py:
```diff
@@ -0,0 +1,102 @@
+
+
+@pytest.mark.cli
+def test_build_application_parser():
```
Is it possible/ worth checking whether the parser was constructed with the expected arguments?

rgreinho replied:
It is, but then we would just be retesting the stdlib (which is already well tested - https://github.com/python/cpython/blob/3.9/Lib/test/test_argparse.py)

jdkandersson replied:
The intent wasn't to test argparse, just whether all of the expected `subparsers.add_parser` and `build_parser.add_argument` were called. The reason for it is because what if someone accidentally comments out/ removes one of the lines? Or they refactor and forget that an argument was supposed to be there? It would be good to check that all the expected arguments were added and, if not, that a test or something else fails. The counter argument would be that it should be checked in a pull request or as a part of an integration test.

jdkandersson commented on tests/test_cli.py:
```diff
@@ -0,0 +1,102 @@
+    mocker.patch("open_alchemy.cli.build_json")
+    args = mock.MagicMock()
+    args.specfile = pathlib.Path("spec.json")
+    cli.build(args)
```
Is it worth checking that the various extensions are passed to the correct functions and that the function gets called with the correct arguments?

rgreinho replied:
Since `args`, is a `MagicMock` object, I don't think it makes too much sense to validate the arguments, but I can absolutely add an `assert_called_once()` check on the mocked functions.

jdkandersson commented on tests/test_cli.py:
```diff
@@ -0,0 +1,102 @@
+    """
+    os.chdir(tmp_path)
+    sys.argv = ["openalchemy"] + command.split()
+    cli.main()
```
Is it worth checking that something happened? after the command was executed?

rgreinho replied:
I'll think about how I can add assert values in my params.

This test is more to ensure that the command ran without raising an error.

jdkandersson commented on tests/test_cli.py:
```diff
@@ -0,0 +1,102 @@
+    WHEN the CLI is called
+    THEN the program runs
+    """
+    os.chdir(tmp_path)
```
Does this impact the state for other tests or is this reset after the test executes?

rgreinho replied:
This is a good question. I would say no because `tmp_path` is temporary and the tests are scoped, but I'll definitely research that and provide a more "verified" answer.

rgreinho replied:
I added `print(os.getcwd()` in all cli tests, then ran `pytest -x -s -m cli` and it gave this output:

```
tests/test_cli.py::test_build_application_parser /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_validate_specfile_invalid_extension /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_validate_specfile_valid[valid yml specfile] /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_validate_specfile_does_not_exist /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_validate_specfile_valid[valid json specfile] /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_build /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_main[cli build all] /private/var/folders/6g/038mjwy91b7_m7zvb9_r_hjw0000gp/T/pytest-of-rgreinhofer/pytest-41/test_main_cli_build_all_0
PASSED
tests/test_cli.py::test_validate_specfile_valid[valid yaml specfile] /private/var/folders/6g/038mjwy91b7_m7zvb9_r_hjw0000gp/T/pytest-of-rgreinhofer/pytest-41/test_main_cli_build_all_0
PASSED
```

I believe this proves that the `os.chdir()` function only impact the current test. 👍

jdkandersson replied:
Good stuff :smile:

jdkandersson commented:
Regarding the tests, it looks like there are quite a few flake8 checks that are failing. I could also spot 2 test cases that failed.

rgreinho commented:
> > I believe I addressed all your comments. Except for the documentation.
> > The README starts to get really long. What would you think of created a dedicated page for the CLI? Or maybe giving a small example in the README, but the having an "Advance usage page" (something more or less like https://www.scrapd.org/usage.html)?
> > We could also use an extension which generates parts of the CLI documentation based on the argparse: https://sphinx-argparse.readthedocs.io/en/latest/.
> > Let me know what your thoughts are regarding this part and I'll start working on the documentation.
> 
> I think it would just be a feature listed in the readme, it doesn't need an example there.
> 
> I think that should be ok, I'm curios to see what they look like. It is important that the docs can still be generated locally (using `make html`) and it would be good to know if there are any impacts to the current read the docs integration.

It looks like none of the available extensions for automatically creating argparse documentation work, so I will abandon this idea 😢 .

jdkandersson commented:
My apologies, I intended to review this today but I got stuck at work until ~10 pm ☹️. I will try to do this tomorrow.

rgreinho commented:
No worries. I still have to fix the tests/linter checks. I'll probably do that over the week end.

rgreinho commented:
Did you change some version or configuration of `flake8` or `pytest` by any chance?

When running `pytest` locally I keep getting this error:

```
---------- coverage: platform darwin, python 3.9.0-final-0 -----------
Name                  Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------
examples/app/api.py      32     32     10      0     0%   3-49
-----------------------------------------------------------------
TOTAL                  4362     32   1586      0    99%

97 files skipped due to complete coverage.
Coverage XML written to file coverage.xml

FAIL Required test coverage of 100.0% not reached. Total coverage: 99.29%
==================================================================== short test summary info =====================================================================
FAILED examples/app/models_autogenerated.py::FLAKE8
FAILED tests/open_alchemy/models_file/test_integration.py::test_generate_type_return[relationship] - assert 1 == 0
FAILED tests/open_alchemy/models_file/test_integration.py::test_generate_type_return[simple] - assert 1 == 0
ERROR tests/examples/app/test_app.py::test_delete_id_hit - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_models_autogen_from_dict[not required not given] - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_models_autogen_from_dict[not required given] - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_delete_id_miss - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_get_id_hit - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_models_autogen_init[not required not given] - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_post - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_get_id_miss - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_models_autogen_init[not required given] - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_patch_id_hit - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_get - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_post_duplicate - ModuleNotFoundError: No module named 'api'
ERROR tests/examples/app/test_app.py::test_patch_id_miss - ModuleNotFoundError: No module named 'api'
=============================================== 3 failed, 2828 passed, 325 skipped, 1 warning, 13 errors in 39.68s ===============================================
```

It seems that the `examples/app/models_autogenerated.py` file is somehow being regenerated when the tests run and is incorrectly formatted.

For the other errors I am not exactly sure where they come from.

When running in the CI (https://github.com/jdkandersson/OpenAlchemy/runs/1303815410?check_suite_focus=true), I get:

```
=========================== short test summary info ============================
FAILED open_alchemy/helpers/__init__.py::FLAKE8
FAILED open_alchemy/column_factory/__init__.py::FLAKE8
FAILED examples/app/models_autogenerated.py::FLAKE8
FAILED open_alchemy/schemas/validation/helpers/__init__.py::FLAKE8
FAILED open_alchemy/models_file/artifacts/__init__.py::FLAKE8
============ 5 failed, 3164 passed, 14 warnings in 73.71s (0:01:13) ============
```
But I haven't modified any of these files.

I did rebase my branch on top of master.

That would be great if you would have some pointers to help me get over these errors 🙏 . In meantime, I'll keep digging 😀

jdkandersson commented:
> Did you change some version or configuration of `flake8` or `pytest` by any chance?
> 
> When running `pytest` locally I keep getting this error:
> 
> ```
> ---------- coverage: platform darwin, python 3.9.0-final-0 -----------
> Name                  Stmts   Miss Branch BrPart  Cover   Missing
> -----------------------------------------------------------------
> examples/app/api.py      32     32     10      0     0%   3-49
> -----------------------------------------------------------------
> TOTAL                  4362     32   1586      0    99%
> 
> 97 files skipped due to complete coverage.
> Coverage XML written to file coverage.xml
> 
> FAIL Required test coverage of 100.0% not reached. Total coverage: 99.29%
> ==================================================================== short test summary info =====================================================================
> FAILED examples/app/models_autogenerated.py::FLAKE8
> FAILED tests/open_alchemy/models_file/test_integration.py::test_generate_type_return[relationship] - assert 1 == 0
> FAILED tests/open_alchemy/models_file/test_integration.py::test_generate_type_return[simple] - assert 1 == 0
> ERROR tests/examples/app/test_app.py::test_delete_id_hit - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_models_autogen_from_dict[not required not given] - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_models_autogen_from_dict[not required given] - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_delete_id_miss - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_get_id_hit - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_models_autogen_init[not required not given] - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_post - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_get_id_miss - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_models_autogen_init[not required given] - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_patch_id_hit - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_get - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_post_duplicate - ModuleNotFoundError: No module named 'api'
> ERROR tests/examples/app/test_app.py::test_patch_id_miss - ModuleNotFoundError: No module named 'api'
> =============================================== 3 failed, 2828 passed, 325 skipped, 1 warning, 13 errors in 39.68s ===============================================
> ```
> 
> It seems that the `examples/app/models_autogenerated.py` file is somehow being regenerated when the tests run and is incorrectly formatted.
> 
> For the other errors I am not exactly sure where they come from.
> 
> When running in the CI (https://github.com/jdkandersson/OpenAlchemy/runs/1303815410?check_suite_focus=true), I get:
> 
> ```
> =========================== short test summary info ============================
> FAILED open_alchemy/helpers/__init__.py::FLAKE8
> FAILED open_alchemy/column_factory/__init__.py::FLAKE8
> FAILED examples/app/models_autogenerated.py::FLAKE8
> FAILED open_alchemy/schemas/validation/helpers/__init__.py::FLAKE8
> FAILED open_alchemy/models_file/artifacts/__init__.py::FLAKE8
> ============ 5 failed, 3164 passed, 14 warnings in 73.71s (0:01:13) ============
> ```
> 
> But I haven't modified any of these files.
> 
> I did rebase my branch on top of master.
> 
> That would be great if you would have some pointers to help me get over these errors 🙏 . In meantime, I'll keep digging 😀

The `models_autogenerated.py` file for the app tests does get re-generated on each test run, it just needs black applied to it and it should be good to go. I have been meaning to automate that, the pre-commit hooks should be taking care of it.

Something weird is going on with the CI for this PR, it does look like it is using a different version of flake8. The F401 errors should be disabled for `__init__.py` files but, for some reason, in this PR they are not. Maybe you need to do something similar as you did for `flake8-max-line-length = 88` but for the following flake8 settings:
```
per-file-ignores =
    */__init__.py:F401
    */models_autogenerated.py:E501
    *models_auto.py:E501
```

rgreinho commented:
Ok, first finding, `flake8` and `pytest-flake8` don't use the same configuration section in the configuration file AND don't use the same configuration format (https://github.com/tholo/pytest-flake8#configuring-flake8-options-per-project-and-file).

But I still have a lot of weird errors about modules not found 🤔

rgreinho commented:
OK, some progress, I got it right on my laptop:

```
---------- coverage: platform darwin, python 3.9.0-final-0 -----------
Name    Stmts   Miss Branch BrPart  Cover   Missing
---------------------------------------------------
---------------------------------------------------
TOTAL    4362      0   1586      0   100%

98 files skipped due to complete coverage.
Coverage XML written to file coverage.xml

Required test coverage of 100.0% reached. Total coverage: 100.00%
======================================================== 2844 passed, 325 skipped, 14 warnings in 35.17s
```

rgreinho commented:
Alright! Everything was addressed! ✅ 

Since I could not get the CLI documentation to be generated from the code, I wrote some instructions by hand, finding some inspiration in the Docker CLI (https://docs.docker.com/engine/reference/commandline/build/). 📚

jdkandersson commented on docs/source/cli.rst:
```diff
@@ -0,0 +1,56 @@
+Usage
+^^^^^
+
+.. program:: openalchemy
```
I think you just need to add `openalchemy` to the `spell.json` file under `// Library` on line 3.

jdkandersson commented on docs/source/cli.rst:
```diff
@@ -0,0 +1,56 @@
+
+.. program:: openalchemy
+
+.. option:: openalchemy build [OPTIONS] SPECFILE PKG_NAME OUTPUT_DIR
```
I think you just need to add `SPECFILE` to the `cspell.json` file under `// Library` on line 3.

jdkandersson commented on docs/source/cli.rst:
```diff
@@ -0,0 +1,56 @@
+Extended Description
+^^^^^^^^^^^^^^^^^^^^
+
+The ``openalchemy build`` command builds a reusable Python package based off of
```
please use `` :samp:`openalchemy build` `` for code syntax in 1st files :smile:

jdkandersson commented on docs/source/cli.rst:
```diff
@@ -0,0 +1,56 @@
+
+Produces the following artifacts::
+
+  dist
```
❤️

jdkandersson commented on docs/source/cli.rst:
```diff
@@ -0,0 +1,56 @@
+      └── simple.egg-info
+
+By default, a source and a wheel package are built, but this behavior can be
+adjusted by using the ``--format`` option.
```
please use `` :samp:`--format` `` for code syntax in rst files 😄

jdkandersson commented:
> Alright! Everything was addressed! ✅
> 
> Since I could not get the CLI documentation to be generated from the code, I wrote some instructions by hand, finding some inspiration in the Docker CLI (https://docs.docker.com/engine/reference/commandline/build/). 📚

Excellent, so close! I think you just need to add `openalchemy` and `SPECFILE` to the `cspell.json` file and then the CI will be happy :smile:
